// @ferocia-oss/osnap/src/errors.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.createError = exports.ErrorCode = void 0;
/**
 * The different types of errors that can happen.
 */
var ErrorCode;
(function (ErrorCode) {
    ErrorCode[ErrorCode["Unknown"] = 0] = "Unknown";
    ErrorCode[ErrorCode["MissingPlatform"] = 1] = "MissingPlatform";
    ErrorCode[ErrorCode["InvalidPlatform"] = 2] = "InvalidPlatform";
    ErrorCode[ErrorCode["MissingXcrun"] = 3] = "MissingXcrun";
    ErrorCode[ErrorCode["MissingAndroidHome"] = 4] = "MissingAndroidHome";
    ErrorCode[ErrorCode["MissingAndroidAdb"] = 5] = "MissingAndroidAdb";
    ErrorCode[ErrorCode["MissingPerl"] = 6] = "MissingPerl";
    ErrorCode[ErrorCode["NoRunningiOSSimulators"] = 7] = "NoRunningiOSSimulators";
    ErrorCode[ErrorCode["NoRunningAndroidEmulators"] = 8] = "NoRunningAndroidEmulators";
    ErrorCode[ErrorCode["AmbiguousAndroidEmulator"] = 9] = "AmbiguousAndroidEmulator";
    ErrorCode[ErrorCode["AmbiguousiOSSimulator"] = 10] = "AmbiguousiOSSimulator";
    ErrorCode[ErrorCode["MissingAndroidEmulator"] = 11] = "MissingAndroidEmulator";
    ErrorCode[ErrorCode["MissingiOSSimulator"] = 12] = "MissingiOSSimulator";
    ErrorCode[ErrorCode["ScreenshotFail"] = 13] = "ScreenshotFail";
    ErrorCode[ErrorCode["CopyToClipboardFail"] = 14] = "CopyToClipboardFail";
    ErrorCode[ErrorCode["ClipboardPlatformUnsupported"] = 15] = "ClipboardPlatformUnsupported";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
var errorMap = (_a = {},
    _a[ErrorCode.MissingPlatform] = "Platform is required.  Must be 'ios' or 'android'.",
    _a[ErrorCode.InvalidPlatform] = "Unrecognized platform.  'ios' or 'android' are supported.",
    _a[ErrorCode.MissingXcrun] = "Unable to find 'xcrun' on your path.",
    _a[ErrorCode.MissingAndroidHome] = "Unable to find ANDROID_HOME environment variable.",
    _a[ErrorCode.MissingAndroidAdb] = "Unable to find adb.",
    _a[ErrorCode.MissingPerl] = "Unable to find perl which is used to clean up goofy output from adb.",
    _a[ErrorCode.NoRunningiOSSimulators] = "No iOS simulators are running.",
    _a[ErrorCode.NoRunningAndroidEmulators] = "No Android emulators or devices connected.",
    _a[ErrorCode.ScreenshotFail] = "Some horrible happened while taking a screenshot.",
    _a[ErrorCode.CopyToClipboardFail] = "Unable to copy the image bytes to the clipboard. Sorry about that.",
    _a[ErrorCode.ClipboardPlatformUnsupported] = "Unable to copy to the clipboard on this platform. PRs welcome.",
    _a[ErrorCode.AmbiguousAndroidEmulator] = "Multiple Android devices connected, please specify a device with -d.",
    _a[ErrorCode.AmbiguousiOSSimulator] = "Multiple iOS devices connected, please specify a device with -d.",
    _a[ErrorCode.MissingAndroidEmulator] = "The specified Android device is currently not connected.",
    _a[ErrorCode.MissingiOSSimulator] = "The specified iOS device is currently not connected.",
    _a);
/**
 * Creates an OSnapError.
 *
 * @param code The error code.
 * @param details An optional error message.
 */
function createError(code, details) {
    var e = new Error();
    e.message = errorMap[code];
    e.details = details;
    e.code = code;
    return e;
}
exports.createError = createError;
